### PR TITLE
fix(android/engine): Remove LanguageListUtil

### PR DESCRIPTION
**Pull request by darcywong00**
With the KeyboardController() refactor, the keyboard cloud catalog information is no longer used for determining packaging/keyboard info.

This PR removes `LanguageListUtil` and the two usages of  `LanguageListUtil.getKeyboardsInfo()`:

* KMKeyboard.setKeyboard(String packageID, String keyboardID, String languageID) - simplify maintenance by converting this method into a wrapper to setKeyboard() with a longer parameter list.
* KMManager.getKeyboardState()

Note: getKeyboardState() has a TODO to compare with the latest cloud keyboard package version. That will be in a future PR

This also makes KMKeyboard consistent. When setKeyboard fails, the default keyboard (at index 0) is loaded. This is for apps that don't have sil_euro_latin as a default keyboard

### CODE REVIEW

File: android/KMEA/app/src/main/java/com/tavultesoft/kmea/KMKeyboard.java [hunk 1]
```diff
@@ -375,96 +375,44 @@ public String specialOSKFontFilename(String filename) {
...
+        k.getKeyboardID(),
+        k.getLanguageID(),
+        k.getKeyboardName(),
+        k.getLanguageName(),
+        k.getFont(),
+        k.getOSKFont());
+    }
+
+    return retVal;
+  }
```
Comment (darcywong00) [inline android/KMEA/app/src/main/java/com/tavultesoft/kmea/KMKeyboard.java:392]: setKeyboard(Keyboard k) was merely moved up here. No change

File: android/KMEA/app/src/main/java/com/tavultesoft/kmea/KMKeyboard.java [hunk 2]
```diff
@@ -375,96 +375,44 @@ public String specialOSKFontFilename(String filename) {
...
-    if (!KMManager.shouldAllowSetKeyboard() || keyboardVersion == null) {
-      Toast.makeText(context, "Can't set " + packageID + "::" + keyboardID + " for " +
-        languageID + " language. Loading default keyboard", Toast.LENGTH_LONG).show();
-      packageID = KMManager.KMDefault_PackageID;
-      keyboardID = KMManager.KMDefault_KeyboardID;
-      languageID = KMManager.KMDefault_LanguageID;
-      retVal = false;
-    }
 
     String kbKey = String.format("%s_%s", languageID, keyboardID);
```
Comment (mcdurdin) [inline android/KMEA/app/src/main/java/com/tavultesoft/kmea/KMKeyboard.java:400]: This should be in a new function as discussed in another comment.

File: android/KMEA/app/src/main/java/com/tavultesoft/kmea/KMKeyboard.java [hunk 3]
```diff
@@ -375,96 +375,44 @@ public String specialOSKFontFilename(String filename) {
...
-    if (dismissHelpBubble()) {
-      Handler handler = new Handler();
-      handler.postDelayed(new Runnable() {
-        @Override
-        public void run() {
-          loadJavascript("showHelpBubble()");
-        }
-      }, 2000);
+    if (!KMManager.shouldAllowSetKeyboard() || kbInfo == null) {
+      Toast.makeText(context, "Can't set " + packageID + "::" + keyboardID + " for " +
```
Comment (mcdurdin) [inline android/KMEA/app/src/main/java/com/tavultesoft/kmea/KMKeyboard.java:408]: Should this be reported to sentry as well?

File: android/KMEA/app/src/main/java/com/tavultesoft/kmea/KMManager.java [hunk 4]
```diff
@@ -1103,34 +1103,36 @@ public static Typeface getKeyboardOskFontTypeface(Context context) {
...
-        // Check version
-        String kbKey = String.format("%s_%s", languageID, keyboardID);
-        HashMap<String, String> kbInfo = keyboardsInfo.get(kbKey);
-        String kbVersion = "1.0";
-        if (kbInfo != null) {
-          kbVersion = kbInfo.get(KMManager.KMKey_KeyboardVersion);
-        }
+    if (kbInfo != null) {
+      // Compare with the cloud package to see if update is available
+      String latestCloudPackageVersion = getLatestKeyboardFileVersion(context, packageID, keyboardID); // fix this
```
[the remark this answers is not in the document]
Comment (darcywong00) [reply android/KMEA/app/src/main/java/com/tavultesoft/kmea/KMManager.java:1132]: This is the part that will get updated in a future PR to compare the package version available from `https://api.keyman.com/package-version`

File: android/KMEA/app/src/main/java/com/tavultesoft/kmea/KMManager.java [hunk 5]
```diff
@@ -1103,34 +1103,36 @@ public static Typeface getKeyboardOskFontTypeface(Context context) {
...
 
-        if (kbVersion != null && (FileUtils.compareVersions(kbVersion, latestVersion) == FileUtils.VERSION_GREATER)) {
-          kbState = KeyboardState.KEYBOARD_STATE_NEEDS_UPDATE;
-        }
+      if (kbVersion != null && (FileUtils.compareVersions(latestCloudPackageVersion, kbVersion) == FileUtils.VERSION_GREATER)) {
```
[the remark this answers is not in the document]
Comment (darcywong00) [reply android/KMEA/app/src/main/java/com/tavultesoft/kmea/KMManager.java:1134]: It sounds like if `kbVersion == null` then the keyboard isn't installed, so `kbState = KeyboardState.KEYBOARD_STATE_NEEDS_DOWNLOAD`. I can add a check after line 1125

>      kbInfo = KeyboardController.getInstance().getKeyboardInfo(index);
>      kbVersion = kbInfo.getVersion();

File: android/KMEA/app/src/main/java/com/tavultesoft/kmea/KMManager.java [hunk 6]
```diff
@@ -1103,34 +1103,36 @@ public static Typeface getKeyboardOskFontTypeface(Context context) {
...
+    if (keyboardID.isEmpty() || languageID.isEmpty()) {
       return kbState;
+    }
 
-    String latestVersion = getLatestKeyboardFileVersion(context, packageID, keyboardID);
-    if (latestVersion == null) {
-      kbState = KeyboardState.KEYBOARD_STATE_NEEDS_DOWNLOAD;
-    } else {
+    // Check latest installed keyboard version
+    String kbKey = String.format("%s_%s", languageID, keyboardID);
```
[the remark this answers is not in the document]
Comment (darcywong00) [reply android/KMEA/app/src/main/java/com/tavultesoft/kmea/KMManager.java]: I added the overload `getKeyboardIndex(languageID, keyboardID)`

File: android/KMEA/app/src/main/java/com/tavultesoft/kmea/KMManager.java [hunk 7]
```diff
@@ -1103,34 +1103,36 @@ public static Typeface getKeyboardOskFontTypeface(Context context) {
...
+    Keyboard kbInfo = null;
+    String kbVersion = "1.0";
```
[the remark this answers is not in the document]
Comment (darcywong00) [reply android/KMEA/app/src/main/java/com/tavultesoft/kmea/KMManager.java]: I've now initialized `kbVersion` to null (keyboard not installed). When the version is still null, the state becomes KEYBOARD_STATE_NEEDS_DOWNLOAD

File: android/KMEA/app/src/main/java/com/tavultesoft/kmea/KMKeyboard.java (hunk 3 above)
Comment (darcywong00) [reply android/KMEA/app/src/main/java/com/tavultesoft/kmea/KMKeyboard.java]: Done in a new method `sendError()` that makes the Toast notification and send the message to Sentry.

File: android/KMEA/app/src/main/java/com/tavultesoft/kmea/KMManager.java [hunk 8]
```diff
@@ -1103,34 +1103,39 @@ public static Typeface getKeyboardOskFontTypeface(Context context) {
...
       kbState = KeyboardState.KEYBOARD_STATE_UP_TO_DATE;
+      kbInfo = KeyboardController.getInstance().getKeyboardInfo(index);
+      kbVersion = kbInfo.getVersion();
+      if (kbVersion == null) {
+        // If version not found, keyboard needs to be downloaded
+        kbState = KeyboardState.KEYBOARD_STATE_NEEDS_DOWNLOAD;
+      }
+    } else {
+      kbState = KeyboardState.KEYBOARD_STATE_NEEDS_DOWNLOAD;
+    }
```
Comment (mcdurdin) [inline android/KMEA/app/src/main/java/com/tavultesoft/kmea/KMManager.java:1131]:
```suggestion
    kbState = KeyboardState.KEYBOARD_STATE_NEEDS_DOWNLOAD;
    if (index != KeyboardController.INDEX_NOT_FOUND) {
      kbInfo = KeyboardController.getInstance().getKeyboardInfo(index);
      kbVersion = kbInfo.getVersion();
      if (kbVersion != null) {
        kbState = KeyboardState.KEYBOARD_STATE_UP_TO_DATE;
      }
    }
```

File: android/KMEA/app/src/main/java/com/tavultesoft/kmea/KMManager.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (mcdurdin) [reply android/KMEA/app/src/main/java/com/tavultesoft/kmea/KMManager.java:1135]: When does this check for updates? Is the result cached?

File: android/KMEA/app/src/main/java/com/tavultesoft/kmea/data/KeyboardController.java [hunk 9]
```diff
@@ -190,6 +190,18 @@ public int getKeyboardIndex(String key) {
...
 
+  /**
+   * Given a languageID and keyboardID, return the index of the matching keyboard.
+   * If no match, returns INDEX_NOT_FOUND
+   * @param languageID - String of the language ID
+   * @param keyboardID - String of the keyboard ID
+   * @return int - Index of the matching keyboard
+   */
+  public int getKeyboardIndex(String languageID, String keyboardID) {
+    String key = String.format("%S_%s", languageID, keyboardID);
```
Comment (mcdurdin) [inline android/KMEA/app/src/main/java/com/tavultesoft/kmea/data/KeyboardController.java:201]:
```suggestion
    String key = String.format("%s_%s", languageID, keyboardID);
```
???

File: android/KMEA/app/src/main/java/com/tavultesoft/kmea/KMManager.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (darcywong00) [reply android/KMEA/app/src/main/java/com/tavultesoft/kmea/KMManager.java:1135]: > When does this check for updates?

`MainActivity` and `SystemKeyboard` call `KMManager.executeResourceUpdate(this);`. Updates are retrieved if the current time is after:
a. an hour after a recently failed attempt
b. a day after the last check

> Is the result cached?

The keyboard versions in the cloud catalog used to be cached in `jsonKeyboardsCache.dat`. A future PR will be caching `package-version` results in `jsonResourcesCache.json`.

File: android/KMEA/app/src/main/java/com/tavultesoft/kmea/KMManager.java [hunk 10]
```diff
@@ -1103,34 +1103,37 @@ public static Typeface getKeyboardOskFontTypeface(Context context) {
...
+    Keyboard kbInfo = null;
+    String kbVersion = null;
+    kbState = KeyboardState.KEYBOARD_STATE_NEEDS_DOWNLOAD;
+    int index = KeyboardController.getInstance().getKeyboardIndex(languageID, keyboardID);
+    if (index != KeyboardController.INDEX_NOT_FOUND) {
+      kbInfo = KeyboardController.getInstance().getKeyboardInfo(index);
+      kbVersion = kbInfo.getVersion();
+      if (kbVersion != null) {
+        // Version exists, so keyboard is up to date
+        kbState = KeyboardState.KEYBOARD_STATE_UP_TO_DATE;
```
Comment (mcdurdin) [inline android/KMEA/app/src/main/java/com/tavultesoft/kmea/KMManager.java:1127]:
```suggestion
      // Compare with the cloud package to see if update is available
      String latestCloudPackageVersion = getLatestKeyboardFileVersion(context, packageID, keyboardID); // fix this
      if (FileUtils.compareVersions(latestCloudPackageVersion, kbVersion) == FileUtils.VERSION_GREATER) {
        kbState = KeyboardState.KEYBOARD_STATE_NEEDS_UPDATE;
      } else {
        kbState = KeyboardState.KEYBOARD_STATE_UP_TO_DATE;
      }
```

File: android/KMEA/app/src/main/java/com/tavultesoft/kmea/KMManager.java [hunk 11]
```diff
@@ -1103,34 +1103,37 @@ public static Typeface getKeyboardOskFontTypeface(Context context) {
...
+      }
+    }
 
-        if (kbVersion != null && (FileUtils.compareVersions(kbVersion, latestVersion) == FileUtils.VERSION_GREATER)) {
-          kbState = KeyboardState.KEYBOARD_STATE_NEEDS_UPDATE;
-        }
+    if (kbInfo != null) {
```
Comment (mcdurdin) [inline android/KMEA/app/src/main/java/com/tavultesoft/kmea/KMManager.java:1131]: One last refactor. I think this block should be deleted and replaced with the block starting in line 1127 as seen in my review comment above.

File: android/KMEA/app/src/main/java/com/tavultesoft/kmea/KMManager.java (hunk 10 above)
Comment (darcywong00) [reply android/KMEA/app/src/main/java/com/tavultesoft/kmea/KMManager.java]: I've incorporated this. With an additional check that `(kbInfo != null)` before getting kbVersion (like 1124).

File: android/KMEA/app/src/main/java/com/tavultesoft/kmea/KMManager.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (darcywong00) [reply android/KMEA/app/src/main/java/com/tavultesoft/kmea/KMManager.java]: In c72b23a8edba0913d7b93ab2de48eeb39c237716, I've also updated LanguageResource and Keyboard to initialize unspecified version to  `null`

**Comment by darcywong00** [comment]
> I am slightly concerned that by now propagating version value null (e.g. LanguageResource.java) in that we may have some missing null checks somewhere?

Maybe next time I'll fallback to version "0.0"

**Pull request merged**